versioned_hdf5/hyperspace.py

```diff
@@ -0,0 +1,363 @@
+# Note: this entire module is compiled by cython with wraparound=False
+# See meson.build for details
+
+from __future__ import annotations
+
+import array
+
+import cython
+import numpy as np
+from cython import Py_ssize_t
+
+_empty_tpl = array.array("l", [])
+_py_ssize_t_nbytes = _empty_tpl.itemsize
+# assert that "long" is the same length as void* on the current platform
+# Note: Py_ssize_t is always the same as np.intp
+assert np.intp().nbytes == _py_ssize_t_nbytes
+
+if cython.compiled:  # pragma: nocover
+    from cython.cimports.cpython import array  # type: ignore
+
+    @cython.ccall
+    @cython.inline
+    def empty_view(n: Py_ssize_t) -> Py_ssize_t[:]:
+        """Functionally the same, but faster, as
+
+        v: Py_ssize_t[:] = np.empty(n, dtype=np.intp)
+
+        Note that this is limited to one dimension.
+        """
+        # array.clone exists only in compiled Cython
+        return array.clone(_empty_tpl, n, zero=False)  # type: ignore[attr-defined]
+
+    @cython.ccall
+    @cython.inline
+    def view_from_tuple(t: tuple[int, ...]) -> Py_ssize_t[:]:
+        """Functionally the same, but faster, as
+
+        v: Py_ssize_t[:] = array.array("l", t)
+        """
+        n = len(t)
+        v: Py_ssize_t[:] = empty_view(n)
+        for i in range(n):
+            v[i] = t[i]
+        return v
+
+else:
+
+    def empty_view(n: Py_ssize_t) -> Py_ssize_t[:]:
+        return array.array("l", b" " * _py_ssize_t_nbytes * n)
+
+    def view_from_tuple(t: tuple[int, ...]) -> Py_ssize_t[:]:
+        return array.array("l", t)
+
+
+@cython.ccall
+def fill_hyperspace(
+    obstacles: Py_ssize_t[:, :],
+    shape: tuple[int, ...],
+) -> Py_ssize_t[:, :]:
+    """Given a N-dimensional space of the given shape and a series of obstacles each one
+    point in size, generate hyperrectangles that cover the whole remaining space
+    without ever crossing an obstacle or another hyperrectangle.
+
+    Parameters
+    ----------
+    obstacles:
+        A list of points to avoid, for example as returned by
+        versioned_hdf5.subchunk_map._modified_chunks_in_selection
```

Comment:
One thing to note here unless I missed it somewhere else - I think these have to be ordered along each dimension, right? This is noted in `_hyperrectangles_between`, but it might be good to include here too.

Reply:
Yes. Amended.